doc/qos/Shared-headroom-pool-design.md

```diff
@@ -0,0 +1,235 @@
+All the buffer configuration is statically calculated and deployed through `CONFIG_DB` without dynamic buffer and there isn't any change regarding the flows. So in this sector we will focus on supporting shared headroom pool on top of dynamic buffer calculation except it is explicitly specified.
+
+#### Config DB Enhancements
+
+##### Table DEFAULT_LOSSLESS_BUFFER_PARAMETER
+
+This table contains the default parameters when generating a ingress buffer profile for ingress lossless traffic. The table has existed in dynamic buffer calculation mode. In this design the `over_subscribe_ratio` is introduced.
+
+###### schema
+
+```schema
+    default_dynamic_th      = 1*2DIGIT              ; Default dynamic_th for dynamically generated buffer profiles
+    over_subscribe_ratio    = 1*3DIGIT              ; Optional. The over subscribe ratio for shared headroom pool. The default value is 1.
+```
+
+This table can be updated on-the-fly.
+
+###### Initialization
+
+Typically all vendors share the identical default RoCE parameters. It should be stored in `/usr/share/sonic/templates/buffers_config.j2` which will be used to render the buffer configuration by `config qos reload`.
+
+***Example***
+
+```json
+    "DEFAULT_LOSSLESS_BUFFER_PARAMETER": {
+        "AZURE": {
+            "default_dynamic_th": "0",
+            "over_subscribe_ratio": "2"
+        }
+    }
+```
+
+#### Algorithm update
+
+We will describe the algorithms via which the headroom information, size of shared buffer pool and size of shared headroom pool is calculated.
+
+##### headroom calculating
+
+###### dynamic headroom calculating
+
+In the dynamic headroom calculating algorithm, the `xon`, `xoff` and `size` are calculated as usual. Originally the `size` equals `xon` + `xoff`. For supporting the shared headroom buffer, the `size` should be updated from `xon` + `xoff` to `xon`.
+
+###### headroom override
+
+For the headroom override mode, the `xon`, `xoff` and `size` are specified by user.
+
+Originally, in the CLI:
+
+- the `size` equals to `xon` + `xoff` if it is not specified. With the shared headroom supported, the `size` equals to `xon` by default. If `size` > `xon`, `size` - `xon` will be treated as `hysteresis`.
+- the `xoff` equals to `size` - `xon` if it is not specified. With the shared headroom supported, the `xoff` must be provided.
+
+##### shared headroom pool size calculating
+
+A new parameter over-subscribe ratio is introduced into `CONFIG_DB` and can be configured through CLI.
+
+Meanwhile, there has been an existing parameter `BUFFER_POOL|ingress_lossless_pool.xoff` which also represents the size of the shared headroom pool. In case both parameters are provided, the `BUFFER_POOL|ingress_lossless_pool.xoff` will take affect and the shared headroom pool size won't be calculated dynamically.
+
+During database migrating, if the `BUFFER_POOL|ingress_lossless_pool.xoff` equals the default value (in the old image), it will be removed and replaced by over-subscribe ratio.
```

Comment:
this is applicable only if we want to be in dynamic mode right?

Reply:
Actually, only MSFT SKUs have SHP configured by default and the dynamic buffer calculation does work on them so this flow won’t be hit.
I’m thinking to remove this sentence.